dateutil/parser.py

```diff
@@ -725,11 +726,20 @@ def _parse(self, timestr, dayfirst=None, yearfirst=None, fuzzy=False,
                     # Token is a number
                     len_li = len(l[i])
 
-                    if (len(ymd) == 3 and len_li in (2, 4) and
-                        res.hour is None and
-                        (i + 1 >= len_l or
-                         (l[i + 1] != ':' and
-                          info.hms(l[i + 1]) is None))):
+                    hms_idx = _find_hms_idx(i, l, info, allow_jump=True)
+                    if hms_idx is not None:
```

Comment:
Just trying to understand the reasoning - why was the `_find_hms_idx` call moved to the beginning here?

Is it because you need to get *and store* the results of `_find_hms_idx` within the `if` call, so you've moved it to the front?

I'm not entirely sure how I feel about this. It may not be hurting anything, but I don't understand the reasoning of the original loop well enough to tell why they are ordered the way they are. At the very least, it seems possible that this will induce some performance hit in some important formats like ISO-8601.

This feels *horribly* un-pythonic to me, but one possible work-around for "need to check *and store* the result" would be some sort of mutable object modified in the function, e.g.:

```python
class MutableReturn(object):
    def __init__(self, val):
        self.return_value = val

def f(x, mr):
    rv = x - 7

    if rv > 2:
        rv = None

    mr.return_value = rv

    return rv

if __name__ == "__main__":
    mr = MutableReturn(None)

    for v in [8, 10]:
        if f(v, mr) is not None:
            print('Value was {}'.format(mr.return_value))
```

Another option that is effectively the same thing but may be saner in this case is to have `_find_hms_idx` cache its last return value as a mutable property, e.g.:

```python
def f(x):
    rv = x - 7

    if rv > 2:
        rv = None

    f.last_return = rv
    return rv

if __name__ == "__main__":
    mr = MutableReturn(None)

    for v in [8, 10]:
        if f(v) is not None:
            print('Value was {}'.format(f.last_return))
```

I think the second one's a bit more elegant, but the first one is thread safe since it doesn't manipulate any globals.

Of course, we'll make `_find_hms_idx` a member function of `parser` once this PR is wrapped up (or as part of it if you are making changes anyway), so we could always use the argument that you can get thread safety by spawning a new `parser` object for each thread - though that's kinda thin to avoid creating a mutable return object.